qiskit_experiments/framework/restless_mixin.py

```diff
@@ -0,0 +1,141 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Restless mixin class."""
+
+from typing import Callable, Sequence
+
+from qiskit.providers import Backend
+from qiskit_experiments.data_processing.data_processor import DataProcessor
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+from qiskit_experiments.data_processing import nodes
+from qiskit_experiments.framework.base_analysis import BaseAnalysis
+
+
+class RestlessMixin:
+    """A mixin to facilitate restless experiments.
+
+    A restless enabled experiment is an experiment that can be run in a restless
+    measurement setting. In restless measurements, the qubit is not reset after
+    each measurement. Instead, the outcome of the previous quantum non-demolition
+    measurement is the initial state for the current circuit. Restless measurements
+    therefore require special data processing which is provided by sub-classes of
+    the :code:`RestlessNode`. Restless experiments are a fast alternative for
+    several calibration and characterization tasks, for details see
+    https://arxiv.org/pdf/2202.06981.pdf.
+    This class makes it possible for users to enter a restless run-mode without having
+    to set all the required run options and the data processor. Furthermore, subclasses
+    can override the :meth:`_get_restless_processor` method if they require more
+    complex restless data processing such as two-qubit calibrations. In addition, this
+    class makes it easy to determine if restless measurements are supported for a given
+    experiments.
+    """
+
+    analysis: BaseAnalysis
+    set_run_options: Callable
+    backend: Backend
+    _physical_qubits: Sequence[int]
+    _num_qubits: int
+
+    def enable_restless(self, rep_delay: float, override_restless_processor: bool = False):
```

Comment:
Experiments are run by calling `backend.run`. https://github.com/Qiskit/qiskit-experiments/blob/d27fe41cec995072fa1db02a14d8437028d81a4e/qiskit_experiments/framework/base_experiment.py#L354
If the backend does not have the run options that restless requires it will raise when `run` is called. I think this is acceptable. Naoki is suggesting that we preempt this by validating the arguments of backend.run. This might be too much overhead, especially since the issue will be caught when run is called. I suggest that instead of doing the validation we extend the doc of the mixin with the required options. This will make it clear what a restless capable backend is.

Reply:
Fair enough. To me this is bit against to the policy of Qiskit which doesn't assume any backend options (indeed rep_delay or esp are not stated in the open pulse/qasm spec). Another option would be making these option names variable (e.g. class attribute) so that user can still update options for their provider. In practice there is no non-IBM provider using experiments, thus current code should be okey.